File: 3/js/src/UnderscorifySubstring.ts
function getLocations(str: string, substr: string): [number, number][] {
	const locations: [number, number][] = [];

	let i = 0;
	while (i < str.length) {
		const begIdx = str.indexOf(substr, i);

		if (begIdx == -1) {
			break;
		}

		locations.push([begIdx, begIdx + substr.length]);
		i = begIdx + 1;
	}

	return locations;
}

function collapse(locations: [number, number][]): [number, number][] {
	if (locations.length <= 1) {
		return locations;
	}

	const collapsedLocations: [number, number][] = [];
	collapsedLocations.push(locations[0]);

	let prev = collapsedLocations[0];

	for (let i = 1; i < locations.length; i++) {
		const cur = locations[i];
		if(cur[0] <= prev[1]) {
			prev[1] = cur[1];
		} else {
			prev = cur;
			collapsedLocations.push(cur);
		}
	}

	return collapsedLocations;
}

function underscorify(str: string, locations: [number, number][]): string {
	let res = str;

	for(let i=locations.length-1; i>=0; i--) {
		const location = locations[i];
		res = res.slice(0, location[0]) + "_" + res.slice(location[0], location[1]) + "_" + res.slice(location[1], res.length);
	}

	return res;
}

function underscorifySubstring(str: string, substr: string) {
	return underscorify(str, collapse(getLocations(str, substr)));
}

(function main() {
	const str = "testthis is a testtest to see if testestest it works";
	const substr = "test";

	console.log(underscorifySubstring(str, substr));
})();

File: 3/java/UnderscorifySubstring.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

class UnderscorifySubtring {
	public static List<Integer[]> getLocations(String string, String substring) {
		List<Integer[]> locations = new ArrayList<>();

		int i = 0;
		while(i < string.length()) {
			int begIdx = string.indexOf(substring, i);

			if(begIdx == -1) {
				break;
			}

			locations.add(new Integer[]{begIdx, begIdx+substring.length()});
			i = begIdx + 1;
		}

		return locations;
	}

	public static List<Integer[]> collapse(List<Integer[]> locations) {
		List<Integer[]> newLocations = new ArrayList<>();

		Integer[] prev = locations.get(0);
		newLocations.add(prev);

		for(int i=1; i<locations.size(); i++) {
			Integer[] cur = locations.get(i);

			if(cur[0] <= prev[1]) {
				prev[1] = cur[1];
			} else {
					newLocations.add(cur);
					prev = cur;
			}
		}

		return newLocations;
	}


	public static String solve(String string, String substring) {
		List<Integer[]> locations = collapse( getLocations(string, substring) );
		StringBuilder res = new StringBuilder(string);

		for(int i=locations.size()-1; i>=0; i--) {
			Integer location[] = locations.get(i);

			res.insert(location[1], "_");
			res.insert(location[0], "_");
		}

		return res.toString();
	}

	public static void main(String args[]) {
		String string = "testthis is a testtest to see if testestest it works";
		String substring = "test";

		System.out.println(solve(string, substring));
	}
}
